### PR TITLE
struct: add KeyFunc

### PR DESCRIPTION
**Pull request by siadat**
The KeyFunc allows us to set the keys for each field dynamically.

By default, KeyFunc returns the value specified by the TagName tag.

The arguments to KeyFunc are the field whose key it returns, and the type of its containing struct.

I need to set the keys dynamically when working with the generated Protocol Buffers structs. I am specifying meta information, including keys, as protobuf [field options](https://developers.google.com/protocol-buffers/docs/proto#customoptions) in the proto files. In my usecase, KeyFunc returns the value of the field option.

### CODE REVIEW

**Comment by siadat** [comment]
It might also be a good idea to pass the value (reflect.Value) of the struct, instead of its type. Hence, the signature for KeyFunc could look like this:
```go
func(reflect.Value, reflect.StructField) string
```
Let me know what you think.

File: structs.go [hunk 1]
```diff
@@ -84,6 +89,8 @@ func (s *Struct) Map() map[string]interface{} {
 	return out
 }
 
+type KeyFunc func(reflect.Type, reflect.StructField) string
```
Comment (fatih) [inline structs.go]: I think this should be named as `TagFunc`. Also it needs a doc comment. 

Second I think we should only pass `reflect.StructField`. It's for naming and I don't want to pass any `reflect.Type` or `reflect.Value` arguments to create any kind of side effects.

File: structs.go [hunk 2]
```diff
@@ -84,6 +89,8 @@ func (s *Struct) Map() map[string]interface{} {
 	return out
 }
 
+type TagFunc func(reflect.StructField) string
```
Comment (fatih) [inline structs.go]: This needs a doc comment as well that describes what this is used for.

File: structs.go [hunk 3]
```diff
@@ -20,16 +20,21 @@ type Struct struct {
 	raw     interface{}
 	value   reflect.Value
 	TagName string
+	TagFunc TagFunc
```
Comment (fatih) [inline structs.go]: This needs a doc comment as well that describes what this is used for.

**Comment by fatih** [comment]
Hi,

Unfortunately, I'm archiving this project and will no longer maintaining it. For more information, please read my blog post: [Taking an indefinite sabbatical from my projects.](https://arslan.io/2018/10/09/taking-an-indefinite-sabbatical-from-my-projects/)

Thanks for the feedback and contribution.

**Pull request closed without merging**